File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def maxDepth(self, root: Optional[TreeNode]) -> int:
#         if not root:
#             return 0
        
#         que = collections.deque([root])
#         depth = 0

#         while que:
#             size = len(que)
#             for _ in range(size):
#                 cur = que.popleft()
#                 if cur.left:
#                     que.append(cur.left)
                
#                 if cur.right:
#                     que.append(cur.right)
                
#             depth += 1
        
#         return depth

# class Solution:
#     def maxDepth(self, root: Optional[TreeNode]) -> int:
#         if not root:
#             return 0
        
#         que = collections.deque([root])
#         depth = 0

#         while que:
#             size = len(que)
#             for _ in range(size):
#                 cur = que.popleft()
#                 if cur.left:
#                     que.append(cur.left)
                
#                 if cur.right:
#                     que.append(cur.right)
                
#             depth += 1
        
#         return depth

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        self.depth = 0;
        self.res = 0;
        self.traverse(root);
        return self.res
    
    def traverse(self, root):
        if not root:
            return
        
        self.depth += 1
        self.res = max(self.res, self.depth)
        self.traverse(root.left)
        self.traverse(root.right)
        self.depth -= 1

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return dfs(root, -1000000000);
    }

    private int dfs(TreeNode root, int maxVal) {
        if (root == null) return 0;
        int res = 0;
        if (root.val >= maxVal) {
            res += 1;
            maxVal = root.val;
        }
        res += dfs(root.left, maxVal) + dfs(root.right, maxVal);
        return res;
    }
}
